examples/reshare/src/dkg/state.rs

```diff
@@ -0,0 +1,621 @@
+        self.msgs
+            .append(section, Event::Ack(player.clone(), ack.clone()))
+            .await
+            .expect("should be able to write to msgs");
+        self.msgs
+            .sync(section)
+            .await
+            .expect("should be able to sync msgs");
+
+        // Update in-memory cache
+        self.get_or_create_epoch(epoch).acks.insert(player, ack);
+    }
+
+    /// Persists a finalized dealer log.
+    pub async fn append_log(&mut self, epoch: EpochNum, dealer: P, log: DealerLog<V, P>) {
+        // Persist to journal
+        let section = epoch.get();
+        self.msgs
+            .append(section, Event::Log(dealer.clone(), log.clone()))
+            .await
+            .expect("should be able to write to msgs");
+        self.msgs
+            .sync(section)
+            .await
+            .expect("should be able to sync msgs");
+
+        // Update in-memory cache
+        self.get_or_create_epoch(epoch).logs.insert(dealer, log);
+    }
+
+    /// Persists new epoch state, advancing to the next epoch.
+    pub async fn append_epoch(&mut self, state: Epoch<V, P>) {
+        // Persist to journal
+        self.states
+            .append(state.clone())
+            .await
+            .expect("should be able to write to state");
+        self.states
+            .sync()
+            .await
+            .expect("should be able to sync state");
+
+        // Update in-memory state first (clone before moving to journal)
+        let size = self.states.size();
+        let epoch = EpochNum::new(size - 1);
+        self.current = Some((epoch, state));
+    }
+
+    /// Removes all data from epochs older than `min`.
+    pub async fn prune(&mut self, min: EpochNum) {
+        let section = min.get();
+        self.msgs
+            .prune(section)
+            .await
+            .expect("should be able to prune msgs");
+        self.states
+            .prune(section)
+            .await
+            .expect("should be able to prune states");
+
+        // Remove old epoch caches
+        self.epochs.retain(|&epoch, _| epoch >= min);
+    }
+
+    /// Create a Dealer for the given epoch, replaying any stored acks.
+    /// Returns None if we've already submitted a log this epoch.
+    pub fn create_dealer<C: Signer<PublicKey = P>>(
+        &self,
+        epoch: EpochNum,
+        signer: C,
+        round_info: Info<V, P>,
+        share: Option<Share>,
+        rng_seed: Summary,
+    ) -> Option<Dealer<V, C>> {
+        if self.has_submitted_log(epoch, &signer.public_key()) {
+            return None;
+        }
+
+        let (mut crypto_dealer, pub_msg, priv_msgs) = CryptoDealer::start(
+            Transcript::resume(rng_seed).noise(b"dealer-rng"),
+            round_info,
+            signer,
+            share,
+        )
+        .expect("should be able to create dealer");
+
+        // Replay stored acks
+        let mut unsent: BTreeMap<P, DealerPrivMsg> = priv_msgs.into_iter().collect();
```

Comment:
It looks like `CryptoDealer` can just return a `BTreeMap<P, DealerPrivMsg>` rather than `Vec<(P, DealerPrivMsg)>`?